File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Util/List/MyList.java
package Util.List;

import java.util.ArrayList;
import java.util.List;

public class MyList<E> implements MyIList<E> {
    private List<E> list;

    public MyList() {
        this.list = new ArrayList<>();
    }

    private boolean validIndex(int index) {
        return index >=0 && index < this.list.size();
    }

    @Override
    public void add(E element) {
        this.list.add(element);
    }

    @Override
    public boolean remove(Object o) {
        return this.list.remove(o);
    }

    @Override
    public int indexOf(Object o) {
        return this.list.indexOf(o);
    }

    @Override
    public E get(int index) throws ListException {
        if(!validIndex(index))
            throw new ListException("Index is out of bounds!");

        return this.list.get(index);
    }

    @Override
    public E set(int index, E element) throws ListException {
        if(!validIndex(index))
            throw new ListException("Index is out of bounds!");

        return this.list.set(index,element);
    }

    @Override
    public int size() {
        return this.list.size();
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();

        for(E element : this.list) {
            stringBuilder.append(element);
            stringBuilder.append("\r\n");
        }

        return stringBuilder.toString();
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Util/List/MyList.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToyLanguage.Util.List
{
    class MyList<T> : MyIList<T>
    {
        private readonly IList<T> list = new List<T>();

        public int Count => list.Count;

        public void Add(T item)
        {
            list.Add(item);
        }

        public T ElementAt(int index)
        {
            if (index < 0 || index >= list.Count) throw new MyListException("Invalid index!");

            return list.ElementAt(index);
        }

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder();

            foreach(var item in list)
            {
                builder.Append(item);
                builder.Append(System.Environment.NewLine);
            }

            return builder.ToString();
        }
    }
}
